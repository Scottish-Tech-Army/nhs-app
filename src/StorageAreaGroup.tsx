import React from "react";
import "./App.css";
import { Link, useNavigate, useParams } from "react-router-dom";
import Navbar from "./Navbar";
import { getStorageAreaGroup } from "./model/DataDefinitions";
import { ReactComponent as EditIcon } from "./icons/edit.svg";
import { ReactComponent as ArrowLeft } from "./icons/arrow-left.svg";

function StorageAreaGroup() {
  let { storageAreaGroupId } = useParams();
  const navigate = useNavigate();

  let storageAreaGroup = getStorageAreaGroup(storageAreaGroupId!);

  if (!storageAreaGroup) {
    return null;
  }

  return (
    <div className="storage-area-group">
      <header>
        <h1>{storageAreaGroup.name}</h1>
        <button
          type="button"
          className="back"
          aria-label="back"
          onClick={() => navigate("/")}
        >
          <ArrowLeft />
        </button>
      </header>
      <main>
        <div className="scroll">
          {storageAreaGroup.storageAreas.map((storageArea, index) => (
            <Link
              className="single-storage-area"
              key={index}
              to={`/area/${storageArea.storageAreaId}`}
            >
              <EditIcon />
              <div className="display-name">{storageArea.name}</div>
            </Link>
          ))}
        </div>
      </main>
      <footer>
        <Navbar />
      </footer>
    </div>
  );
}

export default StorageAreaGroup;
