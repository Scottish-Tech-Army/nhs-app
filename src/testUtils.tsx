import React from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";

import type { AppStore, RootState } from "./data/store";
import boxContentsReducer, { initialState } from "./data/BoxContentsSlice";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProvider(
  component: React.ReactElement,
  {
    preloadedState = { boxContents: initialState },
    store = configureStore({
      reducer: { boxContents: boxContentsReducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {

  return {
    store,
    ...render(<Provider store={store}>{component}</Provider>, renderOptions),
  };
}
