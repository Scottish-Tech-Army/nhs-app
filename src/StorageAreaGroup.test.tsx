/* eslint-disable testing-library/no-node-access */
import { screen } from "@testing-library/react";
import StorageAreaGroup from "./StorageAreaGroup";
import "@testing-library/jest-dom";
import React from "react";
import { renderWithProvider } from "./testUtils";
import { Route, Routes } from "react-router-dom";

describe("StorageAreaGroup", () => {
  it("renders a storage area group", async () => {
    const { container } = renderWithRoute("transfer-bags");

    expect(screen.getByText("Transfer Bags")).toBeDefined();

    [
      "Transfer Bag 1",
      "Transfer Bag 2",
      "Transfer Bag 3",
      "Transfer Bag 4",
      "Transfer Bag 5",
      "Transfer Bag 6",
    ].forEach((title) => {
      expect(screen.getByText(title)).toBeDefined();
    });

    const missingItemsLink = await screen.findByRole("link", {
      name: "missing-items",
    });
    expect(missingItemsLink).toHaveAttribute("href", "/missing-items");

    expect(container).toMatchSnapshot();
  });

  it("navigates to storage area", async () => {
    const { history, user } = renderWithRoute("transfer-bags");

    expect(screen.getByText("Transfer Bags")).toBeDefined();
    expect(history.location.pathname).toEqual("/areas/transfer-bags");

    await user.click(screen.getByText("Transfer Bag 3"));

    expect(history.location.pathname).toEqual("/area/transfer-bag-3");
  });


  it("can return to previous page", async () => {
    const { user, history } = renderWithRoute("transfer-bags");

    await user.click(screen.getByRole("button", { name: "back" }));

    expect(history.location.pathname).toEqual("/");
  });

  function renderWithRoute(storageAreaGroupId: string) {
    // Add routes to get the contents of useParams populated
    return renderWithProvider(
      <Routes>
        <Route path="areas/:storageAreaGroupId" element={<StorageAreaGroup />} />
        <Route path="*" element={<div>Unknown path</div>} />
      </Routes>,
      {
        initialRoutes: ["/", `/areas/${storageAreaGroupId}`],
      }
    );
  }
});
