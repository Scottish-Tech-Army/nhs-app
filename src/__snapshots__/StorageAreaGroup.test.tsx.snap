// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorageAreaGroup renders a storage area group 1`] = `
<div>
  <div
    class="storage-area-group"
  >
    <header>
      <h1>
        Transfer Bags
      </h1>
      <button
        aria-label="back"
        class="back"
        type="button"
      >
        <svg>
          arrow-left.svg
        </svg>
      </button>
    </header>
    <main>
      <div
        class="scroll"
      >
        <a
          class="single-storage-area"
          href="/area/transfer-bag-1"
        >
          <svg>
            edit.svg
          </svg>
          <div
            class="display-name"
          >
            Transfer Bag 1
          </div>
        </a>
        <a
          class="single-storage-area"
          href="/area/transfer-bag-2"
        >
          <svg>
            edit.svg
          </svg>
          <div
            class="display-name"
          >
            Transfer Bag 2
          </div>
        </a>
        <a
          class="single-storage-area"
          href="/area/transfer-bag-3"
        >
          <svg>
            edit.svg
          </svg>
          <div
            class="display-name"
          >
            Transfer Bag 3
          </div>
        </a>
        <a
          class="single-storage-area"
          href="/area/transfer-bag-4"
        >
          <svg>
            edit.svg
          </svg>
          <div
            class="display-name"
          >
            Transfer Bag 4
          </div>
        </a>
        <a
          class="single-storage-area"
          href="/area/transfer-bag-5"
        >
          <svg>
            edit.svg
          </svg>
          <div
            class="display-name"
          >
            Transfer Bag 5
          </div>
        </a>
        <a
          class="single-storage-area"
          href="/area/transfer-bag-6"
        >
          <svg>
            edit.svg
          </svg>
          <div
            class="display-name"
          >
            Transfer Bag 6
          </div>
        </a>
      </div>
    </main>
    <footer>
      <nav>
        <a
          aria-label="directory"
          class="navicon"
          href="/"
        >
          <svg>
            home.svg
          </svg>
          <div>
            Directory
          </div>
        </a>
        <a
          aria-label="missing-items"
          class="navicon"
          href="/missing-items"
        >
          <svg>
            checklist.svg
          </svg>
          <div>
            Missing Items
          </div>
        </a>
        <div
          aria-label="log out"
          class="navicon"
          id="auth-signout"
        >
          <svg>
            logout.svg
          </svg>
          <div>
            Log Out
          </div>
        </div>
      </nav>
    </footer>
  </div>
</div>
`;
